src/app/api/collections/tree.py

```diff
@@ -1,44 +1,95 @@
-from __future__ import (  # Needed for recursive type annotation, can be dropped with Python>3.10
-    annotations,
-)
-
 from uuid import UUID
 
 from aiohttp import ClientSession
-from pydantic import BaseModel
+from elasticsearch_dsl.response import Response
+from glom import Coalesce, Iter, glom
+
+from app.api.collections.models import (
+    Collection,
+    CollectionAttribute,
+    CollectionTreeNode,
+)
+from app.api.collections.vocabs import tree_from_vocabs
+from app.core.config import ELASTIC_TOTAL_SIZE
+from app.elastic.dsl import qbool, qterm
+from app.elastic.elastic import add_base_match_filters
+from app.elastic.search import Search
+
+
+def build_portal_tree(
+    collections: list, root_noderef_id: UUID
+) -> list[CollectionTreeNode]:
+    lut = {str(root_noderef_id): []}
+
+    for collection in collections:
+        if collection.title:
+            portal_node = CollectionTreeNode(
+                noderef_id=collection.noderef_id,
+                title=collection.title,
+                children=[],
+            )
+
+            try:
+                lut[str(collection.parent_id)].append(portal_node)
+            except KeyError:
+                lut[str(collection.parent_id)] = [portal_node]
+
+            lut[str(collection.noderef_id)] = portal_node.children
+
+    return lut.get(str(root_noderef_id), [])
 
-from app.core.constants import COLLECTION_ROOT_ID
 
+def tree_query(node_id: UUID) -> Search:
+    s = add_base_match_filters(
+        Search().query(qbool(filter=qterm(qfield="path", value=node_id)))
+    )
+    s = s.source(
+        ["nodeRef.id", "properties.cm:title", "collections.path", "parentRef.id"]
+    ).sort("fullpath")[:ELASTIC_TOTAL_SIZE]
+    return s
 
-class CollectionTreeNode(BaseModel):
-    noderef_id: UUID
-    title: str
-    children: list[CollectionTreeNode]
 
+def hits_to_collection(hits: Response) -> list[Collection]:
+    collections = []
+    for hit in hits:
+        entry = hit.to_dict()
+        spec = {
+            "title": Coalesce(CollectionAttribute.TITLE.path, default=None),
```

Comment:
I am not a big fan of this glom thing, but maybe I just can't see the benefits.

Either way, I would move the `CollectionAttribute` defintion into this file - or even within this function. I think it is only used here to define the spec. All together, to me this still looks way more complicated than necessary :-/

Reply:
I do agree, this adds some complexity but also makes it more descriptive, since I do not have to check all types. I have no strong opinion in favour of either, i.e., keeping or dropping glom. So as you prefer.

Reply:
I would not move CollectionAttribute here, because it is also used in elastic.py BUT i was able to remove a duplicate and bring everything to models - less code! :D

Reply:
ah, ok :)